Added GroupBy option for actionAdminProductsListingFieldsModifier hook

astterx opened the pull request:
<!-- Thank you for contributing to the PrestaShop project! 

Please take the time to edit the "Answers" rows with the necessary information: -->

| Questions     | Answers
| ------------- | -------------------------------------------------------
| Branch?       | develop
| Description?  | Added group by for sql query to be used with actionAdminProductsListingFieldsModifier hook
| Type?         | improvement
| Category?     | CO
| BC breaks?    | no
| Deprecations? | no
| Fixed ticket? | (optional) If this PR fixes a [Forge](http://forge.prestashop.com/) ticket, please add its complete Forge URL.
| How to test?  | Please indicate how to best verify that this PR is correct.

<!-- Click the form's "Preview button" to make sure the table is functional in GitHub. Thank you! -->

#### Important guidelines

* Make sure [your local branch is up to date](https://help.github.com/articles/syncing-a-fork/) before commiting your changes!
* Your code MUST respect [our Coding Standards](http://doc.prestashop.com/display/PS16/Coding+Standards) (for code written in PHP, JavaScript, HTML/CSS/Smarty/Twig, SQL)!
* Your commit name MUST respect our [naming convention](http://doc.prestashop.com/display/PS16/How+to+write+a+commit+message)!

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/prestashop/prestashop/8514)
<!-- Reviewable:end -->

astterx committed:
Added GroupBy option for actionAdminProductsListingFieldsModifier hook

kpodemski commented:
Hey @astterx 

you can't just additional parameter inside the function before the previous ones as this is breaking change

astterx committed:
Merge branch 'develop' into develop

Quetzacoalt91 commented on src/Adapter/Product/AdminProductDataProvider.php:
```diff
@@ -322,13 +325,14 @@ public function getCatalogProductList($offset, $limit, $orderBy, $sortOrder, $po
-        $total = Db::getInstance()->executeS('SELECT FOUND_ROWS();', true, false);
+        $sql = $this->compileSqlQuery($sqlSelect, $sqlTable, $sqlWhere, $sqlGroupBy, $sqlOrder, $sqlLimit);
+        $products = \Db::getInstance()->executeS($sql, true, false);
+        $total = \Db::getInstance()->executeS('SELECT FOUND_ROWS();', true, false);
```
No need to replace `Db` with `\Db`, as we declared the class at the top of the file with 

```php
use Db;
```

Quetzacoalt91 commented:
Do you have a use case to share with that PR @astterx?

astterx commented:
For a shop with car parts, I created a module to assign every product to a compatible car. A product can be compatible to multiple cars so using hookActionAdminProductsListingFieldsModifier to retrieve also car model for each product, method getCatalogProductList() will return duplicated products, because now is not possible to group by.

astterx committed:
Update AdminProductDataProvider.php

astterx commented:
Hi @kpodemski,

> Hey @astterx
> 
> you can't just additional parameter inside the function before the previous ones as this is breaking change

Method is used only once, just in location where I modified.

Quetzacoalt91 commented on src/Adapter/Admin/AbstractAdminQueryBuilder.php:
```diff
@@ -166,6 +167,11 @@ protected function compileSqlQuery(array $select, array $table, array $where = a
 
+        // GROUP BY
+        if (!empty($groupBy)) {
+            $sql[] = 'GROUP BY '. implode(', ', $groupBy) . PHP_EOL;
```
I feel we may probably handle user data here, and we need to add some sanitization here with

```php
array_map('pSQL', $groupBy);
```

Can you please replace this line with this one:
```php
$sql[] = 'GROUP BY '. implode(', ', array_map('pSQL', $groupBy)) . PHP_EOL;
```

<!-- ping @rGaillard -->

Quetzacoalt91 commented on src/Adapter/Product/AdminProductDataProvider.php:
```diff
@@ -286,12 +286,15 @@ public function getCatalogProductList($offset, $limit, $orderBy, $sortOrder, $po
             $sqlOrder = array('id_product ASC');
         }
 
+        $sqlGroupBy = [];
```
Currently, we initialize arrays with the long syntax form. `[]` should be `array()`

Quetzacoalt91 reviewed (changes requested):
Just some changes to bring and we should be good.

About the method signature, I decided to not ask for anything as we deal with a protected one. Overrides are also not available for this part of the code, so broken code should not occur here.

astterx committed:
Update AdminProductDataProvider.php

astterx committed:
Update AbstractAdminQueryBuilder.php

Quetzacoalt91 commented:
Thank you @astterx

mickaelandrieu commented:
@mbadrani what do you think about this improvement? Can we accept it?

Thanks for your contribution @astterx

astterx committed:
Update sql structure

For a shop with 10K products, search index will stop working correct. Unsigned INT max value will be reached.

astterx commented:
Hi @Quetzacoalt91 ,

Any update?

Quetzacoalt91 commented on install-dev/data/db_structure.sql:
```diff
@@ -1835,15 +1835,15 @@ CREATE TABLE `PREFIX_search_engine` (
 /* List of words available for a given shop & lang */
 CREATE TABLE `PREFIX_search_word` (
-  `id_word` int(10) unsigned NOT NULL auto_increment,
+  `id_word` bigint(20) unsigned NOT NULL auto_increment,
```
New changes were brought to this PR but are unrelated to the original subject. :(

Quetzacoalt91 commented:
Hello @astterx, we were waiting for the QA team before further action.
In the meantime, would you mind removing the additional commit not related to the groupBy parameter?

Thank you

astterx committed:
Update db_structure.sql

astterx commented:
I removed additional commit.

Quetzacoalt91 reviewed (dismissed):
Thank you. Now, this is in the hands of the QA team. ;)

astterx committed:
Update db_structure.sql

For a shop with 10K products, search index will stop working correct. Unsigned INT max value will be reached.

marionf commented:
Hello @astterx & @Quetzacoalt91 
What are the functional tests to do ?

astterx committed:
Merge pull request #1 from astterx/change-id-word-to-bigint

Update db_structure.sql

astterx committed:
Update db_structure.sql

astterx commented:
Hi @marionf 
You should do a regression testing on admin product list.

astterx commented:
Hi. Any update here?

marionf commented:
ping @Quetzacoalt91 can you review it before I test ?

Quetzacoalt91 reviewed (approved):
Still fine for me.

mickaelandrieu commented:
Thanks @astterx !
